pkg/stash/with_etcd.go

```diff
@@ -0,0 +1,95 @@
+package stash
+
+import (
+	"context"
+	"time"
+
+	"github.com/gomods/athens/pkg/config"
+	"github.com/gomods/athens/pkg/errors"
+	"github.com/gomods/athens/pkg/observ"
+	"go.etcd.io/etcd/clientv3"
+	"go.etcd.io/etcd/clientv3/concurrency"
+	"golang.org/x/sync/errgroup"
+)
+
+// WithEtcd returns a distributed singleflight
+// using an etcd cluster. If it cannot connect,
+// then it will panic
+func WithEtcd(endpoints []string) (Wrapper, error) {
+	const op errors.Op = "stash.WithEtcd"
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	defer cancel()
+	c, err := clientv3.New(clientv3.Config{
+		Endpoints:   endpoints,
+		DialTimeout: time.Second * 5,
+	})
+	if err != nil {
+		return nil, errors.E(op, err)
+	}
+	var eg errgroup.Group
+	for _, ep := range endpoints {
+		eg.Go(func() error {
+			_, err := c.Status(ctx, ep)
+			return err
+		})
+	}
+	err = eg.Wait()
+	if err != nil {
+		return nil, errors.E(op, err)
+	}
+	return func(s Stasher) Stasher {
+		return &etcd{c, s}
+	}, nil
+}
+
+type etcd struct {
+	c *clientv3.Client
+	s Stasher
+}
+
+func (s *etcd) Stash(ctx context.Context, mod, ver string) error {
+	const op errors.Op = "etcd.Stash"
+	ctx, span := observ.StartSpan(ctx, op.String())
+	defer span.End()
+
+	mv := config.FmtModVer(mod, ver)
+	sesh, err := concurrency.NewSession(s.c)
+	if err != nil {
+		return errors.E(op, err)
+	}
+	mu := concurrency.NewMutex(sesh, mv)
+	err = mu.Lock(ctx)
+	if err != nil {
+		return errors.E(op, err)
+	}
+	defer mu.Unlock(ctx)
+	resp, err := s.c.Get(ctx, mv)
+	if err != nil {
+		return errors.E(op, err)
+	}
+	if resp.Count == 0 {
+		lease, err := s.c.Grant(ctx, 15*60)
```

Comment:
can we make this configurable?

Reply:
@michalpristas i think our config file is already hard to understand for users. A lot of internal details they might not care about. In this case, I'd rather do Distributed WaitGroup and put the complexity on "us" rather than the "user".

Reply:
so maybe we need a separate way how to treat these.
I'm usually strongly against hardcoded values, of course, everybody can tweak and build but if you found out in prod that for example 10 is enough you need to tweak, build, potentially sign, deploy + all the fancy approval needed in the corporate environment instead of enabling one simple change using config